### PR TITLE
Add CLICK command for preprocess steps

### PR DESCRIPTION
**Pull request by junchaoxia**
Closing issue #311.

### CODE REVIEW

File: src/aspire/commands/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+              help='batch size to perform pipeline so that not crash memory')
+@click.option('--save_mode', default='single',
+              help='Option to save mrc file, if not single, saved to multiply files in batch size')
+@click.option('--overwrite', default=False,
+              help='Option to overwrite output mrc files')
+def preprocess(data_folder, starfile_in, starfile_out, pixel_size, max_rows,
+               flip_phase, max_resolution, normalize_background, whiten_noise,
+               invert_contrast, batch_size, save_mode, overwrite):
+    """
+    Preprocess the ram images and output desired images for future analysis
```
Comment (vineetbansal) [inline src/aspire/commands/preprocess.py]: raw

File: src/aspire/commands/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+              help='Normalize images to background noise or not')
+@click.option('--whiten_noise', default=True,
+              help='whiten background noise')
+@click.option('--invert_contrast', default=True,
+              help='Invert the contrast of images so molecules are shown in white')
+@click.option('--batch_size', default=512,
+              help='batch size to perform pipeline so that not crash memory')
+@click.option('--save_mode', default='single',
+              help='Option to save mrc file, if not single, saved to multiply files in batch size')
+@click.option('--overwrite', default=False,
```
Comment (vineetbansal) [inline src/aspire/commands/preprocess.py]: Whether to overwrite mrc files if they already exist

File: src/aspire/commands/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--max_resolution', default=16, type=int,
+              help='Resolution of downsampled images read from starfile')
+@click.option('--normalize_background', default=True,
+              help='Normalize images to background noise or not')
+@click.option('--whiten_noise', default=True,
+              help='whiten background noise')
+@click.option('--invert_contrast', default=True,
+              help='Invert the contrast of images so molecules are shown in white')
+@click.option('--batch_size', default=512,
+              help='batch size to perform pipeline so that not crash memory')
```
Comment (vineetbansal) [inline src/aspire/commands/preprocess.py]: Batch size to use to save images to .mrc files.

File: src/aspire/commands/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--max_rows', default=None, type=int,
+              help='Max. no. of image rows to read from starfile')
+@click.option('--flip_phase', default=True,
+              help='perform phase flip or not')
+@click.option('--max_resolution', default=16, type=int,
+              help='Resolution of downsampled images read from starfile')
+@click.option('--normalize_background', default=True,
+              help='Normalize images to background noise or not')
+@click.option('--whiten_noise', default=True,
+              help='whiten background noise')
```
Comment (vineetbansal) [inline src/aspire/commands/preprocess.py]: Whiten

File: src/aspire/commands/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+              help='Path to output starfile relative to data folder')
+@click.option('--pixel_size', default=1, type=float,
+              help='Pixel size of images in starfile')
+@click.option('--max_rows', default=None, type=int,
+              help='Max. no. of image rows to read from starfile')
+@click.option('--flip_phase', default=True,
+              help='perform phase flip or not')
+@click.option('--max_resolution', default=16, type=int,
+              help='Resolution of downsampled images read from starfile')
+@click.option('--normalize_background', default=True,
```
Comment (vineetbansal) [inline src/aspire/commands/preprocess.py]: Whether to normalize images to background noise

File: src/aspire/commands/preprocess.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in commit 0b58b0d.

File: src/aspire/commands/preprocess.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in commit 0b58b0d.

File: src/aspire/commands/preprocess.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in commit 0b58b0d.

File: src/aspire/commands/preprocess.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in commit 0b58b0d.

File: src/aspire/commands/preprocess.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in commit 0b58b0d.

File: src/aspire/commands/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--data_folder', default=None,
+              help='Path to data folder')
+@click.option('--starfile_in', required=True,
+              help='Path to input starfile relative to data folder')
+@click.option('--starfile_out', required=True,
+              help='Path to output starfile relative to data folder')
+@click.option('--pixel_size', default=1, type=float,
+              help='Pixel size of images in starfile')
+@click.option('--max_rows', default=None, type=int,
+              help='Max. no. of image rows to read from starfile')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: “Max number”

File: src/aspire/commands/preprocess.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--flip_phase', default=True,
+              help='perform phase flip or not')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: Capitalize.

File: src/aspire/commands/preprocess.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--max_resolution', default=16, type=int,
+              help='Resolution of downsampled images read from starfile')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: This suggests that the images are already downsampled, which is not what we want, right?

File: src/aspire/commands/preprocess.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--normalize_background', default=True,
+              help='Whether to normalize images to background noise')
+@click.option('--whiten_noise', default=True,
+              help='Whiten background noise')
+@click.option('--invert_contrast', default=True,
+              help='Invert the contrast of images so molecules are shown in white')
+@click.option('--batch_size', default=512,
+              help='Batch size to load images from .mrc files.')
+@click.option('--save_mode', default='single',
+              help='Option to save mrc file, if not single, saved to multiply files in batch size')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: “multiple”

File: src/aspire/commands/preprocess.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+from aspire.source.relion import RelionSource
+
+logger = logging.getLogger(__name__)
+
+
+@click.command()
+@click.option('--data_folder', default=None,
+              help='Path to data folder')
+@click.option('--starfile_in', required=True,
+              help='Path to input starfile relative to data folder')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: Throughout:
“starfile” → “STAR file“
“mrc file” → “MRC file”

File: src/aspire/commands/preprocess.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+@click.option('--overwrite', default=False,
+              help='Whether to overwrite mrc files if they already exist')
+def preprocess(data_folder, starfile_in, starfile_out, pixel_size, max_rows,
+               flip_phase, max_resolution, normalize_background, whiten_noise,
+               invert_contrast, batch_size, save_mode, overwrite):
+    """
+    Preprocess the raw images and output desired images for future analysis
+    """
+    # Create a source object for 2D images
+    logger.info('This script illustrates how to preprocess experimental CryoEM images')
```
Comment (janden) [inline src/aspire/commands/preprocess.py]: Illustrates?

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/commands/preprocess.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in new commit 65857e4.

File: src/aspire/commands/preprocess.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in new commit 65857e4.

File: src/aspire/commands/preprocess.py (hunk 8 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Changed to "Resolution for downsampling images" in new commit 65857e4.

File: src/aspire/commands/preprocess.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in new commit 65857e4.

File: src/aspire/commands/preprocess.py (hunk 10 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Done in new commit 65857e4.

File: src/aspire/commands/preprocess.py (hunk 11 above)
Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: Changed to "This script preprocesses experimental CryoEM images" in new commit.

**Comment by junchaoxia** [comment]
@janden Rebased on the develop branch.

File: src/aspire/commands/preprocess.py (hunk 11 above)
Comment (janden) [reply src/aspire/commands/preprocess.py]: But why is this here? If the user is running it, don't they know already what it's supposed to do?

Comment (junchaoxia) [reply src/aspire/commands/preprocess.py]: It tends to print out the script  information to screen or file for future reading. This statement was removed in commit 13b2c79.

**Review by janden — APPROVED** [review]

**Pull request merged**